---
title: "Differential Gene Expression Analysis"
subtitle: "Our Patient Data | Targeted"
author: 
  - Mark Edward M. Gonzales^[De La Salle University, Manila, Philippines, gonzales.markedward@gmail.com]
  - Dr. Anish M.S. Shrestha^[De La Salle University, Manila, Philippines, anish.shrestha@dlsu.edu.ph]
output: html_notebook
---

## I. Preliminaries

### Loading libraries

```{r, warning=FALSE, message=FALSE}
library("tidyverse")
library("tibble")
library("msigdbr")
library("ggplot2")
library("TCGAbiolinks")
library("RNAseqQC")
library("DESeq2")
library("ensembldb")
library("purrr")
library("magrittr")
library("vsn")
library("matrixStats")
library("dplyr")
library("grex")
library("tximport")
```

## II. Converting transcript-level to gene-level abundance

Read the quantification files generated by Salmon.

```{r}
SALMON_DIR <- "../data/patient/quantification/salmon/"
samples <- list.files(path=SALMON_DIR, full.names=TRUE)
files <- file.path(samples, "quant.sf")
names(files) <- str_replace(samples, SALMON_DIR, "")
files
```
Load the transcript-to-gene mapping table.

```{r}
tx2gene <- read.delim("../data/patient/tx2gene.gencode.v47.tsv", header=FALSE)
tx2gene
```

Convert the transcript-level to gene-level abundance.

```{r}
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
counts_matrix <- round(txi$counts)
rownames(counts_matrix) <- cleanid(rownames(counts_matrix))
head(counts_matrix)
```
Construct the metadata table.

```{r}
condition <- c()
for (col in colnames(counts_matrix)) {
  if (endsWith(col, "-N")) {
    condition <- c(condition, "normal")
  } else if (endsWith(col, "-T")) {
    condition <- c(condition, "tumor")
  } else {
    condition <- c(condition, "error")
  }
}

meta <- data.frame(condition = factor(condition))
rownames(meta) <- colnames(counts_matrix)

# Each sample should either be "normal" or "tumor"
stopifnot(any(meta == "error") == FALSE)
meta
```

## III. Differential gene expression analysis

References: 

- Official documentation: https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
- Good balance of theory and hands-on: https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html
- Quality control: https://cran.r-project.org/web/packages/RNAseqQC/vignettes/introduction.html

Construct the `DESeqDataSet` object.

```{r}
dds <- DESeqDataSetFromTximport(
  txi,
	colData = meta, 
	design = ~condition
)
```

#### Quality control

Display quality control (QC) plots (refer to https://cran.r-project.org/web/packages/RNAseqQC/vignettes/introduction.html)

- Total sample counts  
  - Total number of counts for each sample
  - We typically expect all samples to have total counts within the same order of magnitude
  
- Library complexity
  - What fraction of counts is taken up by what fraction of genes
  - Samples showing a different library complexity than the rest might be considered low quality
  
- Gene detection
  - Number of detected genes for each sample

```{r}
plot_total_counts(dds)
```

```{r}
plot_library_complexity(dds)
```

```{r}
plot_gene_detection(dds)
```

Perform gene filtering.

```{r}
dds <- filter_genes(dds)
```

Transform the read counts.

From https://chipster.csc.fi/manual/deseq2-transform.html: <br>
You can use the resulting transformed values only for visualization and clustering, not for differential expression analysis which needs raw counts.

```{r}
vsd <- vst(dds)
mean_sd_plot(vsd)
```

Check the clustering of the samples. 

If you encounter the error `Error in loadNamespace(x) : there is no package called 'ComplexHeatmap'`, uncomment and run the following code block:

```{r}
# install.packages("devtools", dependencies = TRUE)
# devtools::install_github("jokergoo/ComplexHeatmap")
```

```{r, fig.width=10,fig.height=15}
set.seed(42)
plot_sample_clustering(vsd, anno_vars = c("condition"), distance = "euclidean")
```

Perform principal component analysis (PCA).

```{r}
plot_pca(vsd, PC_x = 1, PC_y = 2, shape_by = "condition")
```


### Regulated Cell Death

We obtain the gene sets from RCDdb: https://pmc.ncbi.nlm.nih.gov/articles/PMC11384979/

```{r}
RCDdb <- "../data/public/rcd-gene-list/RCDdb/"
projects <- c("patient")
```

Write utility functions for filtering the gene sets, performing differential gene expression analysis, and plotting the results.

```{r}
filter_gene_set_and_perform_dgea <- function(genes) {
  rcd <- list()

  for (project in projects) {
    rownames(genes) <- genes$gene_id
    rcd[[project]] <- counts_matrix[rownames(counts_matrix) %in% genes$gene_id, ]
    rcd[[project]] <- rcd[[project]][, rownames(meta)]
  }
  
  dds_rcd <- list()
  res_rcd <- list()

  for (project in projects) {
    print(project)
    print("=============")
    dds <- DESeqDataSetFromMatrix(
      countData = rcd[[project]],
      colData = meta,
      design = ~condition
    )
    dds <- filter_genes(dds)
    dds$condition <- relevel(dds$condition, ref = "normal")
    dds_rcd[[project]] <- DESeq(dds)
    res_rcd[[project]] <- results(dds_rcd[[project]])
  }

  deseq.bbl.data <- list()

  for (project in projects) {
    deseq.results <- res_rcd[[project]]
    deseq.bbl.data[[project]] <- data.frame(
      row.names = rownames(deseq.results),
      baseMean = deseq.results$baseMean,
      log2FoldChange = deseq.results$log2FoldChange,
      lfcSE = deseq.results$lfcSE,
      stat = deseq.results$stat,
      pvalue = deseq.results$pvalue,
      padj = deseq.results$padj,
      cancer_type = project,
      gene_symbol = genes[rownames(deseq.results), "gene"]
    )
  }

  deseq.bbl.data.combined <- bind_rows(deseq.bbl.data)
  deseq.bbl.data.combined <- dplyr::filter(deseq.bbl.data.combined, abs(log2FoldChange) >= 1.5 & padj < 0.05)

  return(deseq.bbl.data.combined)
}
```

```{r}
plot_dgea <- function(deseq.bbl.data.combined) {
  sizes <- c("<10^-15" = 4, "10^-10" = 3, "10^-5" = 2, "0.05" = 1)

  deseq.bbl.data.combined <- deseq.bbl.data.combined %>%
    mutate(fdr_category = cut(padj,
      breaks = c(-Inf, 1e-15, 1e-10, 1e-5, 0.05),
      labels = c("<10^-15", "10^-10", "10^-5", "0.05"),
      right = FALSE
    ))

  top_genes <- deseq.bbl.data.combined %>%
    group_by(cancer_type) %>%
    mutate(rank = rank(-abs(log2FoldChange))) %>%
    dplyr::filter(rank <= 10) %>%
    ungroup()

  ggplot(top_genes, aes(y = cancer_type, x = gene_symbol, size = fdr_category, fill = log2FoldChange)) +
    geom_point(alpha = 0.5, shape = 21, color = "black") +
    scale_size_manual(values = sizes) +
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", limits = c(min(deseq.bbl.data.combined$log2FoldChange), max(deseq.bbl.data.combined$log2FoldChange))) +
    theme_minimal() +
    theme(
      axis.text.x = element_text(size = 9, angle = 90, hjust = 1)
    ) +
    theme(legend.position = "bottom") +
    theme(legend.position = "bottom") +
    labs(size = "Adjusted p-value", fill = "log2 FC", y = "Cancer type", x = "Gene")
}
```

#### 1. Necroptosis

Fetch the gene set of interest.

```{r}
genes <- read.csv(paste0(RCDdb, "Necroptosis.csv"))
genes$gene_id <- cleanid(genes$gene_id)
genes <- distinct(genes, gene_id, .keep_all = TRUE)
genes <- subset(genes, gene_id != "")
genes
```

```{r}
deseq.bbl.data.combined <- filter_gene_set_and_perform_dgea(genes)
deseq.bbl.data.combined
```

```{r}
plot_dgea(deseq.bbl.data.combined)
```

#### 2. Ferroptosis

Fetch the gene set of interest.

```{r}
genes <- read.csv(paste0(RCDdb, "Ferroptosis.csv"))
genes$gene_id <- cleanid(genes$gene_id)
genes <- distinct(genes, gene_id, .keep_all = TRUE)
genes <- subset(genes, gene_id != "")
genes
```

```{r}
deseq.bbl.data.combined <- filter_gene_set_and_perform_dgea(genes)
deseq.bbl.data.combined
```

```{r}
plot_dgea(deseq.bbl.data.combined)
```


#### 3. Pyroptosis

Fetch the gene set of interest.

```{r}
genes <- read.csv(paste0(RCDdb, "Pyroptosis.csv"))
genes$gene_id <- cleanid(genes$gene_id)
genes <- distinct(genes, gene_id, .keep_all = TRUE)
genes <- subset(genes, gene_id != "")
genes
```

```{r}
deseq.bbl.data.combined <- filter_gene_set_and_perform_dgea(genes)
deseq.bbl.data.combined
```

```{r}
plot_dgea(deseq.bbl.data.combined)
```