---
title: "Differential Gene Expression Analysis"
subtitle: "Our Patient Data | Untargeted"
author: 
  - Mark Edward M. Gonzales^[De La Salle University, Manila, Philippines, gonzales.markedward@gmail.com]
  - Dr. Anish M.S. Shrestha^[De La Salle University, Manila, Philippines, anish.shrestha@dlsu.edu.ph]
output: html_notebook
---

## I. Preliminaries

### Loading libraries

```{r, warning=FALSE, message=FALSE}
library("tidyverse")
library("tibble")
library("msigdbr")
library("ggplot2")
library("TCGAbiolinks")
library("RNAseqQC")
library("DESeq2")
library("ensembldb")
library("purrr")
library("magrittr")
library("vsn")
library("matrixStats")
library("dplyr")
library("grex")
library("tximport")
library("EnhancedVolcano")
```

## II. Converting transcript-level to gene-level abundance

Read the quantification files generated by Salmon.

```{r}
SALMON_DIR <- "../data/patient/quantification/salmon/"
samples <- list.files(path=SALMON_DIR, full.names=TRUE)
files <- file.path(samples, "quant.sf")
names(files) <- str_replace(samples, SALMON_DIR, "")
files
```
Load the transcript-to-gene mapping table.

```{r}
tx2gene <- read.delim("../data/patient/tx2gene.gencode.v47.tsv", header=FALSE)
names(tx2gene) <- c("Transcript", "Gene", "Symbol")
tx2gene
```

Convert the transcript-level to gene-level abundance.

```{r, message=FALSE}
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
counts_matrix <- round(txi$counts)
rownames(counts_matrix) <- cleanid(rownames(counts_matrix))
head(counts_matrix)
```
Construct the metadata table.

```{r}
condition <- c()
for (col in colnames(counts_matrix)) {
  if (endsWith(col, "-N")) {
    condition <- c(condition, "normal")
  } else if (endsWith(col, "-T")) {
    condition <- c(condition, "tumor")
  } else {
    condition <- c(condition, "error")
  }
}

meta <- data.frame(condition = factor(condition))
rownames(meta) <- colnames(counts_matrix)

# Each sample should either be "normal" or "tumor"
stopifnot(any(meta == "error") == FALSE)
meta
```

## III. Differential gene expression analysis

```{r}
projects <- c("patient")
```

Write utility functions for performing differential gene expression analysis and plotting the results.

```{r}
perform_dgea <- function() {
  rcd <- list()

  for (project in projects) {
    rownames(genes) <- genes$gene_id
    rcd[[project]] <- counts_matrix
    rcd[[project]] <- rcd[[project]][, rownames(meta)]
  }
  
  dds_rcd <- list()
  res_rcd <- list()

  for (project in projects) {
    print(project)
    print("=============")
    dds <- DESeqDataSetFromMatrix(
      countData = rcd[[project]],
      colData = meta,
      design = ~condition
    )
    dds <- filter_genes(dds)
    dds$condition <- relevel(dds$condition, ref = "normal")
    dds_rcd[[project]] <- DESeq(dds)
    res_rcd[[project]] <- results(dds_rcd[[project]])
  }

  deseq.bbl.data <- list()
  
  gene2symbol <- tx2gene[!duplicated(tx2gene$Gene),]
  rownames(gene2symbol) <- gene2symbol$Gene

  for (project in projects) {
    deseq.results <- res_rcd[[project]]
    deseq.bbl.data[[project]] <- data.frame(
      row.names = rownames(deseq.results),
      baseMean = deseq.results$baseMean,
      log2FoldChange = deseq.results$log2FoldChange,
      lfcSE = deseq.results$lfcSE,
      stat = deseq.results$stat,
      pvalue = deseq.results$pvalue,
      padj = deseq.results$padj,
      cancer_type = project,
      gene_symbol = gene2symbol[rownames(deseq.results), "Symbol"]
    )
  }

  deseq.bbl.data.combined <- bind_rows(deseq.bbl.data)
  deseq.bbl.data.combined <- dplyr::filter(deseq.bbl.data.combined, abs(log2FoldChange) >= 1.5 & padj < 0.05)

  return(deseq.bbl.data.combined)
}
```

```{r}
plot_dgea <- function(deseq.bbl.data.combined) {
  EnhancedVolcano(
    deseq.bbl.data.combined, 
    x="log2FoldChange", 
    y="padj", 
    lab=deseq.bbl.data.combined$gene_symbol,
  )
}
```


```{r}
deseq.bbl.data.combined <- perform_dgea()
write.csv(deseq.bbl.data.combined, "../data/patient/differentially-expressed-genes-untargeted.csv")
deseq.bbl.data.combined
```

```{r, fig.width=10,fig.height=15}
plot_dgea(deseq.bbl.data.combined)
```
